### PR TITLE
Video.js player for customized video controls

### LINKED ISSUE #30: Embedded video player
As a user, I'd like to use all the features of a modern video player, when a video is embedded in an OCW page.

For example, in the course page https://ocwnext.odl.mit.edu/courses/18-01sc-single-variable-calculus-fall-2010/sections/1.-differentiation/part-a-definition-and-basic-rules/session-1-introduction-to-derivatives/

#### Designs and Mockups

https://projects.invisionapp.com/d/main#/console/18609637/446978225/preview

#### Acceptance Criteria:

- [ ] Choose a video player and write a brief RFC for team to review. [Video.js](https://videojs.com) seems to work well for OVS
- [x] works with current YouTube videos in course data
- [x] playback bar with playback slider
- [x] play button on left
- [x] mute button on left
- [x] volume slider on left
- [x] current time / total time indicator
- [ ] closed caption indicator on right, showing whether captions are on or off
- [ ] gear menu, with:
   - [ ] playback speed
   - [ ] captions (on/off) 
   - [ ] Quality/resolution picker
- [x] full screen button on right

#### Related Issues

- Single-video page player ([InVision](https://projects.invisionapp.com/d/main#/console/18609637/444868933/preview)) 
- Download video button
- "More" button

#### Out of Scope

- specify video start time using a query string in the URL (similar to OVS, YouTube) 
- Support for playing local video files (in addition to YouTube streams). This would be for mirror drives. 
- Add GTM/GA video event tracking, similar to OVS
- "sticky" caption preference
- support for non-English captions

### PR DESCRIPTION
**Pull request by annagav**
#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [ ] Mobile width screenshots
- [ ] Testing
  - [ ] Code is tested
  - [ ] Changes have been manually tested

#### What are the relevant tickets?
Fix #30

#### What's this PR do?
Adding video.js to play video with our customized video controls.
As of 2018 you cannot remove the the Youtube watch later and share links.
There is a plugin required for the quality selector, but you still need to provide the various quality source files which does not work with youtube.

The things I don't think are working for us:
1) using videojs controls:
- Youtube captions don't move up and down when the control bar appears and disappears ( I spoke to Peter, we are not going to use YT's captions)
- the youtube links , like 'share' and 'watch later' on the top of the video, and the YouTube logo on the bottom right corner are still visible.
- cannot use the gear menu

2) Using Youtube controls: 
- Works great, can set which controls appear in the control bar. But it looks no different from the current embedded video.
- does not match the desired styles.

#### How should this be manually tested?
Run a course, and go to the video lectures section.

This is a screenshot when video js controls are enbaled:
Screen Shot 2021-05-11 at 10 12 02 PM

Here the videojs controls are disabled and YT controls are on.:
Screen Shot 2021-05-13 at 2 55 44 PM

### CODE REVIEW

**Comment by annagav** [comment]
The available videojs-themes don't look great:
Screen Shot 2021-05-14 at 1 45 10 PM

**Comment by annagav** [comment]
New styling:
- moved the progress bar to the top of the control bar
- changed the the time display to show current time/duration
- made the control buttons larger then the default vjs style
- made the control bar non transparent, and now it fully overlays the Youtube logo at the bottom right corner.
Screen Shot 2021-05-17 at 8 47 55 PM

**Comment by abdulkdawson** [comment]
@annagav can we place the progress bar on the same line as the controls. It was a request that the controls should be inline with the rest of the controls.

**Comment by annagav** [comment]
@abdulkdawson 
Screen Shot 2021-05-19 at 12 11 13 PM

**Review by mbertrand — APPROVED** [review]
👍 assuming everyone is happy with the UI

File: course/layouts/partials/youtube_player.html [hunk 1]
```diff
@@ -9,6 +7,6 @@
     width="640" height="385"
     data-setup='{
     "techOrder": ["youtube"],
-    "sources": [{"type": "video/youtube", "src": "https://www.youtube.com/watch?v={{ $youtubeKey }}"}]}'
+    "sources": [{"type": "video/youtube", "src": "https://www.youtube.com/watch?v={{ . }}"}]}'
```
Comment (gumaerc) [inline course/layouts/partials/youtube_player.html:10]: This sets up our `youtube_player.html` partial to accept a single string parameter that is the youtube ID, but we still need a file at `course/layouts/shortcodes/youtube.html` specifically to override the built in `youtube` shortcode to use our own with the following content:

```
{{ $youtubeId := index .Params 0 }}
{{ partial "youtube_player.html" $youtubeId }}
```

Comment (annagav) [reply course/layouts/partials/youtube_player.html:10]: I created it and forgot to add it to the commit 😄

**Review by gumaerc — APPROVED** [review]
👍 LGTM

**Comment by abdulkdawson** [comment]
@annagav I'm sorry for wasting your time. I responded after looking at the wrong version of the mock. You had it right in the previous version. My apologies for the error.

**Comment by pdpinch** [comment]
@annagav does this play the captions from YouTube or from a local file? 

For offline playback, we'll need to play captions from a file instead of from YouTube. But we can address that in another PR, after we have a plan for offline.

**Comment by annagav** [comment]
@abdulkdawson I changed back the styles:
Screen Shot 2021-05-21 at 2 32 21 PM

**Comment by annagav** [comment]
@pdpinch the captions are currently switched off. There is an option to show youtube captions, but I remember you mentioned that we don't want to use them.

**Comment by pdpinch** [comment]
Sorry I wasn't clear. 

We definitely need captions. If we can't link to the captions in the processed course JSON, it would be OK for now to use the YouTube captions.

**Comment by annagav** [comment]
@pdpinch I turned on the youtube captions for the player. @mbertrand I don't know if you would like to look at it again?

**Comment by mbertrand** [comment]
The captions appear but there's no way to turn them off, is that okay?

**Comment by annagav** [comment]
@pdpinch is it okay that the captions can't be switched off by the user for this PR?

**Comment by mbertrand** [comment]
In OVS we use the Youtube controls for YT videos,  videojs native controls for cloudfront videos, because the videojs controls didn't provide controls for toggling captions or quality options from YouTube.  I'm not completely up to speed on what the latest updates to the videojs and the youtube plugin are, but I think it might be necessary to create a custom plugin to add these controls.  https://github.com/videojs/videojs-youtube/issues/564

**Comment by pdpinch** [comment]
It only makes sense to merge this PR if we think we can make video.js work for the requirements on issue #30. If we need to write our own plug-in, that's fine. But I don't want to merge this if we don't think it's moving us in the right direction.

**Comment by annagav** [comment]
@pdpinch @mbertrand is this good to merge then?

**Comment by pdpinch** [comment]
Is there an easy way that I can preview it on netlify? If not, I say merge it and then I'll review on RC -- but I don't want to release until I've had a chance to try it out.

**Comment by annagav** [comment]
@pdpinch I asked Carey, there is currently no way to preview a course on netlify.

**Pull request merged**